### PR TITLE
fix: correct Classic wallpaper black/white conversion algorithm

### LINKED ISSUE #12923: [BUG]: RGB/RGBA Images Render as Black When Setting Hardware Wallet Wallpaper
**Describe the bug**

When prompted to load an image to change the hardware wallet's wallpaper, importing a colored image in RGB/RGBA format results in the image being rendered completely black.

Interestingly, the same image works correctly when it is first converted from RGB to Grayscale in Photoshop and its curves/contrast are adjusted before importing it into the application.

**To Reproduce**

1. Open the hardware wallet wallpaper customization flow.
2. Select a colored image in RGB/RGBA format.
3. Import and load the image.
4. Apply the image as the hardware wallet's wallpaper.
5. Observe that the resulting wallpaper is rendered completely black.

**Expected behavior**

The application should correctly process RGB/RGBA images and convert them into the appropriate image format required by the hardware wallet.

Users should not need to manually convert their images to grayscale using external software before importing them.

**Smartphone (please complete the following information):**

* Device: iPhone
* OS: iOS

**Additional context**

The issue appears to be related to how RGB/RGBA images are being interpreted or converted during the wallpaper processing pipeline.

As a potential fix, the application could preprocess supported images by automatically converting RGB/RGBA input into a normalized grayscale representation before passing it to the hardware wallet's image-processing pipeline.

A possible processing flow would be:

`RGB/RGBA → Normalize Color/Alpha → Grayscale → Tone/Curve Adjustment → Hardware Wallet Image Format`

It may also be useful to provide basic image adjustment controls, such as brightness, contrast, curves, or black/white point adjustment, so users can control how the image will appear on the hardware wallet before it is processed.

This would make the wallpaper workflow more robust and remove the need for users to manually preprocess images in Photoshop or other image-editing software.

### PR DESCRIPTION
**Pull request by ByteZhang1024**
Composite over white before reading pixels, use weighted luminance instead of a plain RGB average, and pick the threshold per image via Otsu's method instead of a fixed 128 cutoff.
fix #12923

### CODE REVIEW

**Comment by sidmorizon** [comment]
@codex review

**Comment by sidmorizon** [comment]
@codex security review

File: packages/shared/src/utils/imageUtils.ts [hunk 1]
```diff
@@ -42,6 +42,43 @@ const range = (length: number) => [...Array(length).keys()];
...
+      weightForeground *
+      (meanBackground - meanForeground) *
+      (meanBackground - meanForeground);
+    if (betweenClassVariance > maxVariance) {
+      maxVariance = betweenClassVariance;
+      threshold = t;
+    }
+  }
+  return threshold;
+}
```
Comment (sidmorizon) [inline packages/shared/src/utils/imageUtils.ts:80]: **New image threshold logic ships without any accompanying unit test**

Severity: `non-severe`

The new per-image brightness cutoff calculation (`otsuThreshold` at `packages/shared/src/utils/imageUtils.ts:46-80`) is added without a regression test, even though the repository's mandatory self-testing checklist requires new/modified logic and every bug fix to come with tests.
Impact: A future change can silently break wallpaper black/white conversion because nothing verifies the new behavior.

<details>
<summary>Repo rule: mandatory unit test / regression test for changed logic</summary>

`.cursor/rules/self-testing.mdc` Section 6 requires: "New/modified logic has corresponding `.test.ts` file", "Tests cover happy path AND edge cases (empty, null, error)", and "Bug fix includes a regression test". This PR adds a new pure function (`otsuThreshold`) plus changed thresholding/inversion logic in `convertToBlackAndWhiteImageBase64` (`packages/shared/src/utils/imageUtils.ts:147-186`) with no test file; sibling utilities in `packages/shared/src/utils/` follow the `*.test.ts` convention (e.g. `avatarUtils.test.ts`). The pure `otsuThreshold` function is trivially testable (uniform image, bimodal histogram, empty input).

</details>

File: packages/shared/src/utils/imageUtils.ts [hunk 2]
```diff
@@ -100,29 +137,46 @@ function convertToBlackAndWhiteImageBase64(
...
-        const bw = avg > 128 ? 255 : 0;
-        // const bw = avg > 128 ? 0 : 255;
         data[i] = bw;
         data[i + 1] = bw;
         data[i + 2] = bw;
       }
 
       // reverse color if white part is more than half
-      if (whiteCount > data.length / 2) {
+      if (whiteCount > pixelCount / 2) {
```
Comment (sidmorizon) [inline packages/shared/src/utils/imageUtils.ts:179]: # 🟠 P1 Otsu 让「过半反色」启发式退化成掷硬币，#12923 的全黑症状很可能仍复现

## 问题

Otsu 的目标函数是 `weightBackground * weightForeground * (meanBackground - meanForeground) ** 2`。对于普通照片这种宽单峰直方图，`(μ₀-μ₁)²` 在阈值附近变化平缓，主导项是 `ω₀·ω₁`，而它在 `ω₀ = ω₁ = 0.5` 处取最大。也就是说 Otsu 输出的阈值天然趋近于亮度中位数，二值化后白像素比例被推到 `0.5` 附近。

下面这行判据恰好卡在同一个点上：

```ts
if (whiteCount > pixelCount / 2) { /* 整图反色 */ }
```

旧代码用固定 `128`，白像素占比真实反映图像明暗，只有中位亮度恰好接近 `128` 的图才会踩到临界点；换成 Otsu 之后，几乎每张照片的 `whiteCount` 都落在 `pixelCount / 2` 附近。

## 影响

同一张图裁剪位置、`compress` 系数或 JPEG 量化的微小差异，就足以让 `whiteCount` 跨过 `pixelCount / 2`，壁纸整体在「正片」和「负片」之间随机翻转，用户重试一次得到相反结果。

更关键的是 #12923 报告的「彩色图导入后整片全黑」本身就是这条反色启发式造成的：一张偏亮的照片过半像素被判为白，随后整图取反 → 屏幕上是大面积黑底。本 PR 没有改动这段逻辑，Otsu 之后有约一半概率仍然命中反色分支，issue 描述的症状会原样重现，只是变得不可预测。

## 建议

把反色决策与阈值解耦，不要再用二值化后的白像素占比。可行方向之一是改用原始亮度的统计量（例如 `luminance` 的均值/中位数是否高于中灰）来判断是否需要反色；或者干脆确认硬件端的期望极性后固定极性、去掉这条启发式，并加上明确的边界值（避免 `whiteCount` 恰好等于 `pixelCount / 2` 时的抖动）。无论哪种，都建议对「亮照片」补一条回归用例，确认输出不再是大面积黑底。

[Claude session ↗](https://claude.ai/code/session_01Spi7vAhoes6BPHHEAyZTzu)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/shared/src/utils/imageUtils.ts [hunk 3]
```diff
@@ -100,29 +137,46 @@ function convertToBlackAndWhiteImageBase64(
...
+      const luminance = new Uint8ClampedArray(pixelCount);
+      for (let i = 0, p = 0; i < data.length; i += 4, p += 1) {
+        luminance[p] = toGrayScale(data[i], data[i + 1], data[i + 2]);
+      }
+
+      // Otsu threshold instead of a fixed 128 — adapts per image without a
+      // stretch, which a couple of outlier pixels could otherwise skew.
+      let threshold = 128;
+      try {
+        threshold = otsuThreshold(luminance);
```
Comment (sidmorizon) [inline packages/shared/src/utils/imageUtils.ts:159]: # 🟠 P1 Otsu 缺少可分性下限，低对比度图会被放大成 JPEG 噪点雪花

## 问题

`otsuThreshold` 无条件返回类间方差最大的那个 `t`，没有判断图像是否真的存在两个可分的亮度簇。只要图像里存在 **两个及以上** 不同亮度值，循环就会命中 `betweenClassVariance > maxVariance` 并把 `threshold` 落在噪声带内部；只有像素完全同值时才会保留 `threshold = 128` 的默认值。

调用点上游是 `manipulateAsync(uri, actions, { compress: compress || 0.8, format: SaveFormat.JPEG, base64: true })`，也就是说进来的一定是 JPEG。纯色/低对比度区域经过 DCT 量化后不是严格同值，而是在几个相邻灰阶上抖动（例如 `28`~`33`）。此时 Otsu 会在这个 5 级宽的噪声带里挑一个阈值，把块间量化误差直接放大成硬边二值。

## 影响

一张接近纯色或低对比度的壁纸（纯色背景、柔和渐变、夜景照片），旧逻辑用固定 `128` 输出干净的整片黑，新逻辑会输出按 8x8 JPEG 块随机分布的黑白雪花点，之后再叠加一次可能的整图反色。这是肉眼可见的画质回退，而且越是「用户主观认为很简单」的图越容易触发。

## 建议

在 `otsuThreshold` 里加一个可分性下限，不满足时回退到固定阈值（或直接输出单色）。例如用动态范围或归一化类间方差判断：

```ts
// 只有当两簇确实分离时才采用 Otsu 结果
const separability = maxVariance / (total * total * totalVariance);
if (separability < MIN_SEPARABILITY) return 128;
```

也可以用更简单的 `max(luminance) - min(luminance) < MIN_RANGE`（例如 `32`）作为守卫。另外 `try/catch` 包住 `otsuThreshold` 实际上捕不到任何可达异常，真正需要的兜底是上面这个可分性判断。

[Claude session ↗](https://claude.ai/code/session_01Spi7vAhoes6BPHHEAyZTzu)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/shared/src/utils/imageUtils.ts [hunk 4]
```diff
@@ -100,29 +137,46 @@ function convertToBlackAndWhiteImageBase64(
       canvas.width = img.width;
       canvas.height = img.height;
 
+      // Composite over white so transparent RGBA pixels aren't read as black below.
+      ctx.fillStyle = '#FFFFFF';
+      ctx.fillRect(0, 0, canvas.width, canvas.height);
```
Comment (sidmorizon) [inline packages/shared/src/utils/imageUtils.ts:142]: # 🟡 P2 白底合成在真实调用链上是空操作，alpha 早在上游就被压平了

## 问题

`convertToBlackAndWhiteImageBase64` 在仓库里只有一个实际调用点（同文件 `imageUtils.ts` 的 `isMonochrome` 分支），而它的输入来自紧邻上游的：

```ts
const imageResult: ImageResult = await manipulateAsync(uri, actions, {
  compress: compress || 0.8,
  format: SaveFormat.JPEG,
  base64: true,
});
```

JPEG 不带 alpha 通道。等这段 base64 传进来时，透明像素已经被 `expo-image-manipulator` 压平成不透明的背景色（原生编码器通常压到黑），`img` 上根本不存在 `alpha < 255` 的像素。native 路径经 `getImageEmbedBridge()` → webembed 走的也是同一条链，输入同样是这份 JPEG。

## 影响

新增的 `fillStyle = '#FFFFFF'` + `fillRect` 对 Classic 壁纸这条路径不产生任何像素差异，等于每张图多做一次全画布填充。更重要的是，issue #12923 里 RGBA 透明区域变黑的问题并没有被修掉——真正压平 alpha 的位置在 `manipulateAsync` 那一步，这里补白底已经太晚，后续如果有人依赖「透明会被当成白」的语义会得到错误结论。

## 建议

要真正处理 RGBA，需要把白底合成前移到 `manipulateAsync` 之前（或让该步骤输出 `SaveFormat.PNG` 保留 alpha，由本函数负责合成）。如果只是想保留一层防御，建议把注释改成说明「上游已压平 alpha，此处仅为防御」，避免误导后续维护者以为透明输入已经被正确处理。

[Claude session ↗](https://claude.ai/code/session_01Spi7vAhoes6BPHHEAyZTzu)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by originalix — CHANGES_REQUESTED** [review]
合到 x 分支吧，不在 6.5.2 单独测了，qa 来不及
